File: packages/core/src/ledger.ts
export interface LedgerEntry {
  readonly path: string;
  readonly sha256: string;
  readonly mtime_ms: number;
  readonly size_bytes: number;
  readonly lines_returned: number;
  readonly offset: number;
  readonly limit: number;
  readonly timestamp_ms: number;
}

export interface Ledger {
  record(entry: LedgerEntry): void;
  getLatest(path: string): LedgerEntry | undefined;
  getAll(path: string): readonly LedgerEntry[];
  clear(): void;
}

export class InMemoryLedger implements Ledger {
  private readonly entries = new Map<string, LedgerEntry[]>();

  record(entry: LedgerEntry): void {
    const list = this.entries.get(entry.path) ?? [];
    list.push(entry);
    this.entries.set(entry.path, list);
  }

  getLatest(p: string): LedgerEntry | undefined {
    const list = this.entries.get(p);
    if (!list || list.length === 0) return undefined;
    return list[list.length - 1];
  }

  getAll(p: string): readonly LedgerEntry[] {
    return this.entries.get(p) ?? [];
  }

  clear(): void {
    this.entries.clear();
  }
}

File: crates/write/src/ledger.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::sync::{Arc, Mutex};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LedgerEntry {
    pub path: String,
    pub sha256: String,
    pub mtime_ms: u64,
    pub size_bytes: u64,
    pub timestamp_ms: u64,
}

/// Mirror of the TS `Ledger` surface: keyed by path, most-recent wins.
pub trait Ledger: Send + Sync + std::fmt::Debug {
    fn get_latest(&self, path: &str) -> Option<LedgerEntry>;
    fn record(&self, entry: LedgerEntry);
}

#[derive(Default)]
pub struct InMemoryLedger {
    inner: Arc<Mutex<HashMap<String, LedgerEntry>>>,
}

impl InMemoryLedger {
    pub fn new() -> Self {
        Self {
            inner: Arc::new(Mutex::new(HashMap::new())),
        }
    }
}

impl std::fmt::Debug for InMemoryLedger {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("InMemoryLedger").finish()
    }
}

impl Ledger for InMemoryLedger {
    fn get_latest(&self, path: &str) -> Option<LedgerEntry> {
        self.inner.lock().unwrap().get(path).cloned()
    }
    fn record(&self, entry: LedgerEntry) {
        self.inner
            .lock()
            .unwrap()
            .insert(entry.path.clone(), entry);
    }
}
